app/refreshdata.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+""" Refresh data files for the Covid19 Mtl dashboard """
+
+import os
+import sys
+import fcntl
+import csv
+
+from datetime import datetime
+from argparse import ArgumentParser
+
+import requests
+import lxml
+import bs4
+
+# Data sources mapping
+# {filename: url}
+SOURCES = { #'cases.csv': '', 
+            #'cases_per1000': '', 
+            'data_mtl.html': 'https://santemontreal.qc.ca/population/coronavirus-covid-19/',
+
+            # The main page for this dataset is https://www.inspq.qc.ca/covid-19/donnees 
+            'data_qc.csv': 'https://www.inspq.qc.ca/sites/default/files/covid/donnees/combine.csv'}
+DATA_DIR = 'data'
+NB_RETRIES = 3
+
+def lock(lock_dir):
+    ''' Lock the data directory to prenvent concurent runs of the scraper, 
+    which would be risky for data corruption. '''
+
+    lockf = os.path.join(lock_dir, 'scraper.pid')
+    if not os.path.isfile(lockf):
+
+        open(lockf, 'w') # create the empty file
+    fd = os.open(lockf, os.O_RDONLY)
+    fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+
+    # No unlocking needed.  fcntl() locks are released then the process exits.
+    open(lockf, 'w').write('{}'.format(os.getpid))
+
+
+def fetch(url):
+    ''' Get the data at `url`.  Our data sources are notoriously unreliable, 
+    so we retry a few times. '''
+    for i in range(NB_RETRIES):
+        resp = requests.get(url)
+        if resp.status_code != 200:
+            next
+        return resp.content
+    raise RuntimeError('Failed to retrieve {}'.format(url))
+
+
+def save_df(filename, data):
+    ''' Save a datafile if it's newer and at least as big as what we cached.  
+    Raise ValueError otherwise.'''
+    if os.path.isfile(filename):
+        # TODO: double check that this independent of the encoding
```

Comment:
This is needed because you think the requests will get different encodings over time?

We should probably always save in UTF-8, and for `data`, you could change `fetch` to return the response file object, then we can check if it’s UTF-8 and if not, transcode it using codecs.StreamRecoder.

Reply:
I was trying to avoid reading the actual bytes to save IO in case we have big files, but that's probably the bullet-proof way to do it.  If encoding changes, and anything can happen with the third party data sources, counting bytes is not valid and we have to count characters instead.   

It would be reasonable to decode at read time and to encode to UTF-8, assuming that what's on disk will always be UTF-8.  I guess that's what we should do.

Reply:
Ho!  HTTP header `Content-Length` is the number of bytes, you can use that.